File: src/main/java/no/nav/familie/integrasjoner/personopplysning/PersonopplysningerService.java
package no.nav.familie.integrasjoner.personopplysning;

import no.nav.familie.integrasjoner.client.rest.PdlRestClient;
import no.nav.familie.integrasjoner.client.soap.PersonSoapClient;
import no.nav.familie.integrasjoner.felles.ws.DateUtil;
import no.nav.familie.integrasjoner.personopplysning.domene.PersonhistorikkInfo;
import no.nav.familie.integrasjoner.personopplysning.domene.Personinfo;
import no.nav.familie.integrasjoner.personopplysning.domene.TpsOversetter;
import no.nav.familie.integrasjoner.personopplysning.internal.Person;
import no.nav.tjeneste.virksomhet.person.v3.informasjon.Informasjonsbehov;
import no.nav.tjeneste.virksomhet.person.v3.informasjon.NorskIdent;
import no.nav.tjeneste.virksomhet.person.v3.informasjon.Periode;
import no.nav.tjeneste.virksomhet.person.v3.informasjon.PersonIdent;
import no.nav.tjeneste.virksomhet.person.v3.meldinger.HentPersonRequest;
import no.nav.tjeneste.virksomhet.person.v3.meldinger.HentPersonResponse;
import no.nav.tjeneste.virksomhet.person.v3.meldinger.HentPersonhistorikkRequest;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.cache.annotation.Cacheable;
import org.springframework.stereotype.Service;
import org.springframework.web.context.annotation.ApplicationScope;

import java.time.LocalDate;
import java.util.List;
import java.util.Objects;


@Service
@ApplicationScope
public class PersonopplysningerService {

    private static final Logger LOG = LoggerFactory.getLogger(PersonopplysningerService.class);
    private static final Logger secureLogger = LoggerFactory.getLogger("secureLogger");
    public static final String PERSON = "PERSON";
    private final PdlRestClient pdlRestClient;
    private final PersonSoapClient personSoapClient;
    private TpsOversetter oversetter;

    @Autowired
    public PersonopplysningerService(PersonSoapClient personSoapClient, TpsOversetter oversetter, PdlRestClient pdlRestClient) {
        this.personSoapClient = personSoapClient;
        this.oversetter = oversetter;
        this.pdlRestClient = pdlRestClient;
    }

    PersonhistorikkInfo hentHistorikkFor(String personIdent, LocalDate fom, LocalDate tom) {
        Objects.requireNonNull(personIdent, "personIdent");
        Objects.requireNonNull(fom, "fom");
        Objects.requireNonNull(tom, "tom");

        var request = new HentPersonhistorikkRequest();
        request.setAktoer(new PersonIdent().withIdent(new NorskIdent().withIdent(personIdent)));
        request.setPeriode(new Periode().withFom(DateUtil.convertToXMLGregorianCalendar(fom))
                                        .withTom(DateUtil.convertToXMLGregorianCalendar(tom)));
        var response = personSoapClient.hentPersonhistorikkResponse(request);
        return oversetter.tilPersonhistorikkInfo(
                new no.nav.familie.integrasjoner.personopplysning.domene.PersonIdent(personIdent), response);
    }

    public Personinfo hentPersoninfoFor(String personIdent) {
        HentPersonRequest request = new HentPersonRequest()
                .withAktoer(new PersonIdent().withIdent(new NorskIdent().withIdent(personIdent)))
                .withInformasjonsbehov(List.of(Informasjonsbehov.FAMILIERELASJONER, Informasjonsbehov.ADRESSE));
        HentPersonResponse response = personSoapClient.hentPersonResponse(request);
        return oversetter.tilPersoninfo(new no.nav.familie.integrasjoner.personopplysning.domene.PersonIdent(personIdent),
                                        response);
    }


    @Cacheable(cacheNames = PERSON, key = "#personIdent", condition = "#personIdent != null")
    public Personinfo hentPersoninfo(String personIdent) {
        HentPersonRequest request = new HentPersonRequest()
                .withAktoer(new PersonIdent().withIdent(new NorskIdent().withIdent(personIdent)))
                .withInformasjonsbehov(List.of(Informasjonsbehov.FAMILIERELASJONER, Informasjonsbehov.ADRESSE));
        HentPersonResponse response;

        response = personSoapClient.hentPersonResponse(request);

        return oversetter.tilPersoninfo(new no.nav.familie.integrasjoner.personopplysning.domene.PersonIdent(personIdent),
                                        response);
    }

    public Person hentPersoninfo(String personIdent, String tema) {
        return pdlRestClient.hentPerson(personIdent, tema);
    }
}

File: src/main/java/no/nav/familie/integrasjoner/personopplysning/PersonopplysningerService.kt
package no.nav.familie.integrasjoner.personopplysning

import no.nav.familie.integrasjoner.client.rest.PdlClientCredentialRestClient
import no.nav.familie.integrasjoner.client.rest.PdlRestClient
import no.nav.familie.integrasjoner.client.rest.RegoppslagRestClient
import no.nav.familie.integrasjoner.felles.OppslagException
import no.nav.familie.integrasjoner.personopplysning.internal.ADRESSEBESKYTTELSEGRADERING
import no.nav.familie.integrasjoner.personopplysning.internal.Adressebeskyttelse
import no.nav.familie.integrasjoner.personopplysning.internal.FORELDERBARNRELASJONROLLE
import no.nav.familie.integrasjoner.personopplysning.internal.PdlPersonMedRelasjonerOgAdressebeskyttelse
import no.nav.familie.integrasjoner.personopplysning.internal.Person
import no.nav.familie.integrasjoner.personopplysning.internal.PersonMedAdresseBeskyttelse
import no.nav.familie.integrasjoner.personopplysning.internal.PersonMedRelasjoner
import no.nav.familie.integrasjoner.personopplysning.internal.PostadresseResponse
import no.nav.familie.kontrakter.felles.Tema
import no.nav.familie.kontrakter.felles.personopplysning.FinnPersonidenterResponse
import no.nav.familie.kontrakter.felles.personopplysning.PersonIdentMedHistorikk
import org.springframework.cache.annotation.Cacheable
import org.springframework.stereotype.Service
import org.springframework.web.context.annotation.ApplicationScope

@Service
@ApplicationScope
class PersonopplysningerService(
    private val pdlRestClient: PdlRestClient,
    private val pdlClientCredentialRestClient: PdlClientCredentialRestClient,
    private val regoppslagRestClient: RegoppslagRestClient,
) {
    fun hentPersoninfo(
        personIdent: String,
        tema: Tema,
    ): Person = pdlRestClient.hentPerson(personIdent, tema)

    fun hentIdenter(
        personIdent: String,
        tema: Tema,
        medHistorikk: Boolean,
    ): FinnPersonidenterResponse {
        val response = pdlRestClient.hentIdenter(personIdent, "FOLKEREGISTERIDENT", tema, medHistorikk)
        return FinnPersonidenterResponse(response.map { PersonIdentMedHistorikk(it.ident, it.historisk) })
    }

    @Cacheable(cacheNames = ["PERSON_MED_RELASJONER"], key = "#personIdent + #tema", condition = "#personIdent != null")
    fun hentPersonMedRelasjoner(
        personIdent: String,
        tema: Tema,
    ): PersonMedRelasjoner {
        val hovedperson = hentPersonMedRelasjonerOgAdressebeskyttelse(listOf(personIdent), tema).getOrThrow(personIdent)
        val barnIdenter =
            hovedperson.forelderBarnRelasjon
                .filter { it.relatertPersonsRolle == FORELDERBARNRELASJONROLLE.BARN }
                .mapNotNull { it.relatertPersonsIdent }
        val sivilstandIdenter = hovedperson.sivilstand.mapNotNull { it.relatertVedSivilstand }.distinct()

        val tilknyttedeIdenter = (barnIdenter + sivilstandIdenter).distinct()
        val tilknyttedeIdenterData = hentPersonMedRelasjonerOgAdressebeskyttelse(tilknyttedeIdenter, tema)

        val barnOpplysninger = tilknyttedeIdenterData.filter { (ident, _) -> barnIdenter.contains(ident) }
        val barnsForeldrer = hentBarnsForeldrer(barnOpplysninger, personIdent, tema)

        return PersonMedRelasjoner(
            personIdent = personIdent,
            adressebeskyttelse = hovedperson.gradering(),
            sivilstand = mapPersonMedRelasjoner(sivilstandIdenter, tilknyttedeIdenterData),
            barn = mapPersonMedRelasjoner(barnIdenter, tilknyttedeIdenterData),
            barnsForeldrer = barnsForeldrer,
        )
    }

    fun hentAdressebeskyttelse(
        personIdent: String,
        tema: Tema,
    ): Adressebeskyttelse =
        pdlRestClient.hentAdressebeskyttelse(personIdent, tema).adressebeskyttelse.firstOrNull()
            ?: Adressebeskyttelse(gradering = ADRESSEBESKYTTELSEGRADERING.UGRADERT)

    fun hentPostadresse(
        personIdent: String,
        tema: Tema,
    ): PostadresseResponse? = regoppslagRestClient.hentPostadresse(personIdent, tema)

    private fun hentBarnsForeldrer(
        barnOpplysninger: Map<String, PdlPersonMedRelasjonerOgAdressebeskyttelse>,
        personIdent: String,
        tema: Tema,
    ): List<PersonMedAdresseBeskyttelse> {
        val barnsForeldrerIdenter =
            barnOpplysninger
                .flatMap { (_, personMedRelasjoner) ->
                    personMedRelasjoner.forelderBarnRelasjon.filter { it.relatertPersonsRolle != FORELDERBARNRELASJONROLLE.BARN }
                }.filter { it.relatertPersonsIdent != personIdent }
                .mapNotNull { it.relatertPersonsIdent }
                .distinct()
        val barnsForeldrerOpplysninger = hentPersonMedRelasjonerOgAdressebeskyttelse(barnsForeldrerIdenter, tema)

        return mapPersonMedRelasjoner(barnsForeldrerIdenter, barnsForeldrerOpplysninger)
    }

    private fun mapPersonMedRelasjoner(
        sivilstandIdenter: List<String>,
        tilknyttedeIdenterData: Map<String, PdlPersonMedRelasjonerOgAdressebeskyttelse>,
    ) = sivilstandIdenter.map { PersonMedAdresseBeskyttelse(it, tilknyttedeIdenterData.getOrThrow(it).gradering()) }

    private fun PdlPersonMedRelasjonerOgAdressebeskyttelse.gradering() = this.adressebeskyttelse.firstOrNull()?.gradering

    private fun Map<String, PdlPersonMedRelasjonerOgAdressebeskyttelse>.getOrThrow(ident: String) =
        this[ident] ?: throw OppslagException(
            "Finner ikke $ident i response fra pdl",
            "pdl",
            OppslagException.Level.MEDIUM,
        )

    private fun hentPersonMedRelasjonerOgAdressebeskyttelse(
        personIdenter: List<String>,
        tema: Tema,
    ): Map<String, PdlPersonMedRelasjonerOgAdressebeskyttelse> {
        if (personIdenter.isEmpty()) return emptyMap()
        return pdlClientCredentialRestClient.hentPersonMedRelasjonerOgAdressebeskyttelse(personIdenter, tema)
    }
}
